File: Trees/BSTTraversal.java

import java.util.*;

class Program
{
  // In Order Traversal ( left root right)
  // Traverse the left sub tree : InOrder(leftsubtree)
  // Traverse the right sub tree : InOrder(right subtree)
  // O(n) time | O(n) space
  public static ArrayList<Integer> inOrderTraverse(BST tree, ArrayList<Integer> array) {
    if(tree == null)
		{
			return array;
		}
		else
		{
			inOrderTraverse(tree.left, array);
			array.add(tree.value);
			inOrderTraverse(tree.right, array);
		}
		return array;
  }

  // Pre Order Traversal(root left right)
  // Vist the root
  // Traverse the left subtree : PreOrder(left-subtree)
  // Traverse the right subtree : PreOrder(right-subtree)
  public static ArrayList<Integer> preOrderTraverse(BST tree, ArrayList<Integer> array) {
    array.add(tree.value);
		if(tree.left != null)
		{ preOrderTraverse(tree.left, array);
		}
		if(tree.right != null)
		{
		preOrderTraverse(tree.right, array);
		}
		return array;
  }

  // Post Order Traversal(left right root)
  // Traverse the left subtree : PostOrder(left-subtree)
  // Traverse the right subtree: PostOrder(right-subtree)
  // Visit the root

  public static ArrayList<Integer> postOrderTraverse(BST tree, ArrayList<Integer> array) {
  if(tree.left != null)
  {
  postOrderTraverse(tree.left, array);
  }
  if(tree.right != null)
  {
  postOrderTraverse(tree.right, array);
  }
  array.add(tree.value);
  return array;
}

static class BST {
  public int value;
  public BST left;
  public BST right;

  public BST(int value) {
    this.value = value;
    }
  }
}

File: DailyCodingProblem/C++/BSTTraversal.cpp
#include <vector>
using namespace std;

class BST {
  public:
    int value;
    BST* left;
    BST* right;

    BST(int val);
    BST& insert(int val);
};

//left, root, right
// O(n) time | O(n) space
vector<int> inOrderTraverse(BST* tree, vector<int> array) {

  if(tree->left != NULL)
	{
		array = inOrderTraverse(tree->left, array);
	}
	array.push_back(tree->value);
	if (tree->right != NULL)
	{
		array = inOrderTraverse(tree->right, array);
	}
	return array;
}

//root, left, right
// O(n) time | O(n) space
vector<int> preOrderTraverse(BST* tree, vector<int> array) {
  array.push_back(tree->value);
	if(tree->left != NULL)
	{
		array = preOrderTraverse(tree->left, array);
	}
	if (tree->right != NULL)
	{
		array = preOrderTraverse(tree->right, array);
	}
	return array;
}

//left, right, root
// O(n) time | O(n) space 
vector<int> postOrderTraverse(BST* tree, vector<int> array) {
  if(tree->left != NULL)
	{
		array = postOrderTraverse(tree->left, array);
	}
	if(tree->right != NULL)
	{
		array = postOrderTraverse(tree->right, array);
	}
	array.push_back(tree->value);
	return array;
}
